src/lib/mathlib/math/filter/MedianFilter.hpp

```diff
@@ -0,0 +1,91 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2020 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/*
+ * @file MedianFilter.hpp
+ *
+ * @brief Implementation of a median filter with C qsort.
+ *
+ */
+
+#pragma once
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+namespace math
+{
+
+template<typename T, int WINDOW = 3>
+class MedianFilter
+{
+public:
+	static_assert(WINDOW >= 3, "MedianFilter window size must be >= 3");
+	static_assert(WINDOW % 2, "MedianFilter window size must be odd"); // odd
+
+	MedianFilter() = default;
+
+	void insert(const T &sample)
+	{
+		_head = (_head + 1) % WINDOW;
+		_buffer[_head] = sample;
+	}
+
+	T median()
+	{
+		T sorted[WINDOW];
+		memcpy(sorted, _buffer, sizeof(_buffer));
+		qsort(&sorted, WINDOW, sizeof(T), cmp);
```

Comment:
Maybe a median-of-3 algorithm would work well here? It might be implemented smaller than the qsort (unless qsort is included already, then qsort is find)

Reply:
We already include qsort in parameters (utarray) and gyro calibration. I initially implemented this at a few fixed sizes with a sorting network (a generated set of SWAP macros), but the lazy qsort version seemed good enough.

https://pages.ripco.net/~jgamble/nw.html